packages/react-dom/src/client/ReactDOMComponent.js

```diff
@@ -1110,12 +1110,11 @@ export function diffHydratedProperties(
       } else if (propKey === DANGEROUSLY_SET_INNER_HTML) {
         const serverHTML = domElement.innerHTML;
         const nextHtml = nextProp ? nextProp[HTML] : undefined;
-        const expectedHTML = normalizeHTML(
-          domElement,
-          nextHtml != null ? nextHtml : '',
-        );
-        if (expectedHTML !== serverHTML) {
-          warnForPropDifference(propKey, serverHTML, expectedHTML);
+        if (nextHtml != null) {
```

Comment:
So.. what if `nextHtml` is `null` or `undefined`, but it was not `null` on the server? Wouldn't this miss a warning?

Reply:
I had the exact same question. This is pretty similar to the second test I added (https://github.com/facebook/react/pull/18676/files#diff-ab371863932cd2e8f0ba14ff2eaab380R568). I can change it not render anything on the client or add a new one that goes from `{ __html: "server" }` to `undefined`.

Reply:
Adding new one sounds good.

Reply:
I'm just asking how it works.

Reply:
I didn't read much of the surrounding code and relied on the tests doing what I expected so I can't tell you how it works.

Maybe I should convert it to a table and `it.each` so that it's more visible  how the props change between client and server? The tests are mostly setup.

Reply:
We prefer explicitly written tests to `it.each`. Don't worry about duplication.

>I didn't read much of the surrounding code and relied on the tests doing what I expected so I can't tell you how it works.

I think we need to figure out how it works before we can be comfortable merging this. :-)

Reply:
> I think we need to figure out how it works before we can be comfortable merging this. :-)

Sure thing. Let me trace this for each test case. Maybe this clears this up.

Reply:
When tracing the code I wasn't sure if this might result in a loss of component stacks so I added these to the test to make sure.

It looks like the reconciler is going through all the instances that aren't hydrated and deletes them:
https://github.com/facebook/react/blob/dc630d3374e8b3d01cdf9e12a81eab9322ac6832/packages/react-reconciler/src/ReactFiberHydrationContext.old.js#L462

The host config is then responsible for warning about the deletion of hydrateable text/elements: https://github.com/facebook/react/blob/58c895e59cd2822a9a7a67606d5fe7da5e09ba2a/packages/react-dom/src/client/ReactDOMHostConfig.js#L957

So it's no longer warning about a mismatch in `dangerouslySetInnerHTML` but rather that the resulting html mismatches.

It seems like both approaches do not create a helpful error message. In the version on `master` it displays the wrong content of `dangerouslySetInnerHTML` thinking that it's `<p>server</p>` while this content came from the `children` prop. In the version on this PR we no longer see that `dangerouslySetInnerHTML` was part of the issue.